packages/core/src/actors.ts

```diff
@@ -161,54 +167,72 @@ export function fromCallback<TEvent extends EventObject>(
 export function fromPromise<T>(
   lazyPromise: Lazy<PromiseLike<T>>
 ): Behavior<{ type: string }, T | undefined> {
-  let canceled = false;
-  const resolveEventType = Symbol('resolve');
-  const rejectEventType = Symbol('reject');
+  const resolveEventType = '$$xstate.resolve';
+  const rejectEventType = '$$xstate.reject';
 
   // TODO: add event types
-  const behavior: Behavior<any, T | undefined> = {
-    transition: (state, event, { self, name }) => {
-      if (canceled) {
+  const behavior: Behavior<
+    any,
+    T | undefined,
+    { canceled: boolean; data: T | undefined }
+  > = {
+    transition: (state, event, { self, id, observers }) => {
+      const _event = toSCXMLEvent(event);
+      if (state.canceled) {
         return state;
       }
 
-      switch (event.type) {
-        case startSignalType:
-          const resolvedPromise = Promise.resolve(lazyPromise());
-
-          resolvedPromise.then(
-            (response) => {
-              self.send({ type: resolveEventType, data: response });
-            },
-            (errorData) => {
-              self.send({ type: rejectEventType, data: errorData });
-            }
-          );
-          return undefined;
+      const eventObject = _event.data;
+
+      switch (_event.name) {
         case resolveEventType:
           self._parent?.send(
-            toSCXMLEvent(doneInvoke(name, event.data) as any, {
+            toSCXMLEvent(doneInvoke(id, eventObject.data) as any, {
               origin: self
             })
           );
-          return event.data;
+
+          // Make sure that observers are completed only after the
+          // snapshot has been set on the actor
+          setTimeout(() => {
+            observers.forEach((o) => o.complete?.());
+          });
```

Comment:
Having this here means that the design is not complete and we need to amend this before we can land this PR. Those behaviors here should not care about `observers` at all - and most definitely should not use a `setTimeout` (or any other async function) to call observers. 

I'm tinkering with this locally to see what are our options. So far I have:
- return a `.status` (or similar) from here and recognize the failure state [here](https://github.com/statelyai/xstate/blob/ec4a4c185bbb3ceccd8004d2930a83e0074e845e/packages/core/src/interpreter.ts#L314-L319), just after `.transition` call
- have smth like `fail()` in the actor context

I prefer the first one as that would standardize this across all behaviors and looks cleaner.

The problem is not only about timing but also about "finalizing" the interpreter - marking it as done, clearing all subscriptions etc. The same stuff happens when a machine reaches a final state and the same should happen here. So I guess this could be handled through a `getStatus` here:
https://github.com/statelyai/xstate/blob/ec4a4c185bbb3ceccd8004d2930a83e0074e845e/packages/core/src/interpreter.ts#L187-L190

But we need to differentiate between `done` and `error` "statuses".

Reply:
All the observer calls have now been removed BTW. Next step is to provide stricter typing for the "status object"